src/osd/ECBackend.cc

```diff
@@ -1012,23 +1043,74 @@ void ECBackend::handle_sub_read_reply(
       make_pair(
 	from,
 	i->second));
-    if (rop.complete[i->first].r == 0)
-      rop.complete[i->first].r = i->second;
+    dout(20) << __func__ << " shard=" << from << " error=" << i->second << dendl;
   }
 
   map<pg_shard_t, set<ceph_tid_t> >::iterator siter =
-shard_to_read_map.find(from);
+					shard_to_read_map.find(from);
   assert(siter != shard_to_read_map.end());
   assert(siter->second.count(op.tid));
   siter->second.erase(op.tid);
 
   assert(rop.in_progress.count(from));
   rop.in_progress.erase(from);
-  if (!rop.in_progress.empty()) {
-    dout(10) << __func__ << " readop not complete: " << rop << dendl;
-  } else {
-    dout(10) << __func__ << " readop complete: " << rop << dendl;
+  unsigned is_complete = 0;
+  // For redundant reads check for completion as each shard comes in,
+  // or in a non-recovery read check for completion once all the shards read.
+  if (rop.do_redundant_reads || (!rop.for_recovery && rop.in_progress.empty())) {
+    for (map<hobject_t, read_result_t>::const_iterator iter =
+        rop.complete.begin();
+      iter != rop.complete.end();
+      ++iter) {
+      set<int> have;
+      for (map<pg_shard_t, bufferlist>::const_iterator j =
+          iter->second.returned.front().get<2>().begin();
+        j != iter->second.returned.front().get<2>().end();
+        ++j) {
+        have.insert(j->first.shard);
+        dout(20) << __func__ << " have shard=" << j->first.shard << dendl;
+      }
+      set<int> want_to_read, dummy_minimum;
+      get_want_to_read_shards(&want_to_read);
+      int err;
+      // XXX: Could just do if (have.size < ec_impl->get_data_chunk_count())
```

Comment:
`ec_impl->minimum_to_decode() < 0` is not equivalent to `(have.size < ec_impl->get_data_chunk_count())`. if the former is true, we can tell that the later is also. but not vice versa.